### DISCUSSION TITLE
Failed to convert the python code to C/Fortran: Function is missing a type annotation, and so on.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [pyccel/pyccel#1532] [Q&A]
Hi,

The environment is Ubuntu 20.04 with python 3.8.3 managed by pyenv. I try to convert [this python script](https://github.com/hongyi-zhao/test/blob/master/mu.py) into C/Fortran code with the help of pyccel, but all my tries failed. See following for more info:

```
$ pyccel --language fortran --compiler gfortran -t mu.py 

ERROR at annotation (semantic) stage
pyccel: [semantic] 
 |error: mu.py [11]| Function is missing a type annotation (fermi)
 |fatal: mu.py [12,1]| Undefined variable (E)

$ pyccel --language fortran --compiler ifort -t mu.py 

ERROR at annotation (semantic) stage
pyccel: [semantic] 
 |error: mu.py [11]| Function is missing a type annotation (fermi)
 |fatal: mu.py [12,1]| Undefined variable (E)

$ pyccel --language c --compiler icc -t mu.py 

ERROR at annotation (semantic) stage
pyccel: [semantic] 
 |error: mu.py [11]| Function is missing a type annotation (fermi)
 |fatal: mu.py [12,1]| Undefined variable (E)
```
Any hints for this problem?

Regards,
HY

**Comment by EmilyBourne** [comment]
Hi,

The issue seems to be that you have not told pyccel which types your function `fermi` expects as input. Unfortunately pyccel does not have an up to date documentation to point you to. The data types must be specified for each function argument. This can either be done using the pyccel decorator `types`:
```
from pyccel.decorators import types

@types('real[:]','real','real')
def fermi(E,mu,kT):
```
or using a comment:
```
#$ header function fermi(real[:], real, real)
def fermi(E,mu,kT):
```

I suggest that you look at the file https://github.com/pyccel/pyccel/blob/master/tests/pyccel/scripts/runtest_funcs.py for an example with decorators or https://github.com/pyccel/pyccel/blob/master/tests/codegen/fcode/scripts/functions.py for an example with comments

**Comment by EmilyBourne** [comment]
I have had a quick look at your code, and it seems that you use a few functions that aren't implemented in pyccel:
- np.genfromtxt
- np.argmax
- np.arange (the python equivalent `range` is implemented and seems to be sufficient for your example)
- np.trapz
- np.column_stack
- np.savetxt

Some of these could be implemented easily (e.g. argmax), so don't hesitate to open another issue if you feel something is important. However others (e.g genfromtxt) would be harder. We usually recommend that you put your computing core in a separate file and just accelerate that part (ie. the bottleneck). For example, (using python `range` and after implementation of a `trapz` function) the following section https://github.com/hongyi-zhao/test/blob/master/mu.py#L26-L44 could be accelerated along with the `fermi` function

**Comment by hongyi-zhao** [comment]
1. The core of [the python code](https://github.com/hongyi-zhao/test/blob/master/mu.py) is the *trapz* function. In fact, there are many ready-to-use implementations of it, say,  [stdlib.fortran-lang.org](https://stdlib.fortran-lang.org/interface/trapz.html), [www.rosettacode.org](http://www.rosettacode.org/wiki/Numerical_integration), and so on.
2. In the long run, from the point of view of pyccel itself, even I open another new issue, it doesn't do much to solve the whole problem in nature due that there are so many functions in numpy/scipy, etc., whose conversion rules aren't so easy to implement in pyccel.

**Comment by saidctb** [comment]
>     1. The core of [the python code](https://github.com/hongyi-zhao/test/blob/master/mu.py) is the _trapz_ function. In fact, there are many ready-to-use implementations of it, say,  [stdlib.fortran-lang.org](https://stdlib.fortran-lang.org/interface/trapz.html), [www.rosettacode.org](http://www.rosettacode.org/wiki/Numerical_integration), and so on.
> 
>     2. In the long run, from the point of view of pyccel itself, even I open another new issue, it doesn't do much to solve the whole problem in nature due that there are so many functions in numpy/scipy, etc., whose conversion rules aren't so easy to implement in pyccel.

I get that you want to have access to all numpy functions, but this is not possible and never will be, you will always need to refactor your code and accelerate the computational part only, so something like `np.genfromtxt` and `np.savetxt` is not in our priority and Pyccel's aim is to translate a subset of the pure python code  and support (partially) some libraries like `numpy, scipy, mpi4py, ...`

**Comment by hongyi-zhao** [comment]
> Pyccel's aim is to translate subset of pure python code and support (partially) some libraries like `numpy, scipy, mpi4py, ...`

If so, I think the following ones should have the priority for consideration, as far as the current problem are concerned:
```
np.trapz
np.column_stack
```

**Comment by yguclu** [comment]
@hongyi-zhao
Did you ever have a look at the source code for Numpy's `trapz`? This is a very basic algorithm, and it only takes a few lines of codes to write it down. See [here](https://github.com/numpy/numpy/blob/92ebe1e9a6aeb47a881a1226b08218175776f9ea/numpy/lib/function_base.py#L4006-L4094).

If you simply copy-paste that code to another file, and then simplify it a bit to work on one-dimensional arrays, Pyccel can accelerate it easily.

**Comment by hongyi-zhao** [comment]
@yguclu Thanks for your advice.
